File: labbox_ephys_widgets/widgets/RecordingsView/RecordingsView.py
from labbox_ephys import dbcollection
import labbox_ephys as le
import numpy as np

class RecordingsView:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self._refresh_recordings()

    def on_message(self, msg):
        if msg['action'] == 'remove_recordings':
            recording_ids = msg['recording_ids']
            dbcollection('recordings').remove({'recording_id': {'$in': recording_ids}})
            self._refresh_recordings()
        if msg['action'] == 'refresh_recordings':
            self._refresh_recordings()
        if msg['action'] == 'sort_recordings':
            recording_ids = msg['recording_ids']
            sorter = msg['sorter']
            for recording_id in recording_ids:
                recording = le.load_recording(recording_id=recording_id)
                all_channel_ids = recording.get_channel_ids()
                all_channel_groups = recording.get_channel_groups()
                all_group_ids = list(set(all_channel_groups))
                lookup = dict()
                for i in range(len(all_channel_ids)):
                    lookup[all_channel_ids[i]] = all_channel_groups[i]
                for group_id in all_group_ids:
                    sorting_id = f'{recording_id}::{group_id}::{sorter["name"]}'
                    channel_ids = [ch for ch in all_channel_ids if lookup[ch] == group_id]
                    doc = dict(
                        recording_id=recording_id,
                        group_id=group_id,
                        channel_ids=_list(channel_ids),
                        recording=recording.object(),
                        sorting=None,
                        sorting_name=sorter['name'],
                        sorter_name=sorter['name'],
                        sorting_id=sorting_id,
                        unit_ids=None,
                        runtime_info=None,
                        status='pending'
                    )
                    if dbcollection('sortings').find(dict(sorting_id=sorting_id)).count() == 0:
                        dbcollection('sortings').insert_one(doc)
                    else:
                        print(f'Not inserting sorting because it already exists: {sorting_id}')
            # dbcollection('recordings').remove({'recording_id': {'$in': recording_ids}})
            self._refresh_recordings()
        
    
    def _refresh_recordings(self):
        self._set_status('running', 'Loading recordings...')
        recordings = list(dbcollection('recordings').find(dict()))
        self._set_state(recordings=recordings)
        self._set_status('finished', 'Finished loading recordings.')
    
    # Send a custom message to JavaScript side
    # In .js file, use this.pythonInterface.onMessage((msg) => {...})
    def _send_message(self, msg):
        self.send_message(msg)

    # Set the python state
    def _set_state(self, **kwargs):
        self.set_state(kwargs)
    
    # Set error status with a message
    def _set_error(self, error_message):
        self._set_status('error', error_message)
    
    # Set status and a status message. Use running', 'finished', 'error'
    def _set_status(self, status, status_message=''):
        self._set_state(status=status, status_message=status_message)

def _list(x):
    if isinstance(x, np.ndarray):
        return x.tolist()
    return x

File: labbox_ephys_widgets/widgets/RecordingsView/RecordingsView.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import RecordingsTable from './RecordingsTable.js'
import { Toolbar, IconButton } from '@material-ui/core';
import { FaTrash, FaSync } from 'react-icons/fa';
import Button from '@material-ui/core/Button';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
const config = require('./RecordingsView.json');

const sorter_options = [
    {
        label: 'MountainSort4',
        sorter: {
            name: "mountainsort4",
            parameters: {}
        }
    },
    {
        label: 'KiloSort2',
        sorter: {
            name: "kilosort2",
            parameters: {}
        }
    }
];

export default class RecordingsView extends Component {
    static title = 'View database of recordings'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            
            // python state
            recordings: null,
            status: '',
            status_message: '',

            //
            selectedRecordingIds: {}
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.start();
        this.setState({
            status: 'started',
            status_message: 'Starting python backend'
        });
        this.pythonInterface.setState({
            trigger: 1
        });
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    _handleDeleteSelectedRecordings = () => {
        const recordingIds = this.state.selectedRecordingIds;
        this.pythonInterface.sendMessage({action: 'remove_recordings', recording_ids: recordingIds});
    }
    _handleSortSelectedRecordings = (sorter) => {
        const recordingIds = this.state.selectedRecordingIds;
        this.pythonInterface.sendMessage({action: 'sort_recordings', recording_ids: recordingIds, sorter: sorter});
    }
    _handleRefreshRecordings = () => {
        this.setState({
            recordings: null,
            status: '',
            status_message: ''
        });
        this.pythonInterface.sendMessage({action: 'refresh_recordings'});
    }
    render() {
        const recordings = this.state.recordings;
        const selectedRecordingIds = this.state.selectedRecordingIds;

        let content;
        if (recordings) {
            content = (
                <RecordingsTable
                    recordings={recordings}
                    selectedRecordingIds={this.state.selectedRecordingIds}
                    onSelectedRecordingIdsChanged={(ids) => {this.setState({selectedRecordingIds: ids})}}
                />
            );
        }
        else {
            content = (
                <ReportStatus {...this.state} />
            );
        }

        return (
            <div>
                <Toolbar>
                    <IconButton title={"Refresh recordings"} onClick={() => {this._handleRefreshRecordings()}}>
                        <FaSync />
                    </IconButton>
                    <IconButton disabled={isEmpty(selectedRecordingIds)} title={"Delete selected recordings"} onClick={() => {this._handleDeleteSelectedRecordings()}}>
                        <FaTrash />
                    </IconButton>
                    <DropdownMenu
                        label="Launch spike sorting"
                        disabled={isEmpty(selectedRecordingIds)} 
                        options={sorter_options}
                        onSorterSelected={(sorter) => {this._handleSortSelectedRecordings(sorter)}}
                    />
                </Toolbar>
                <div style={{height: 250, overflowY: 'auto'}}>
                    {content}
                </div>
            </div>
        );
    }
}

class ReportStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'started':
                return <div>Started: {this.props.status_message}</div>;
            case 'running':
                return <div>{this.props.status_message}</div>;
            case 'error':
                return <div>Error: {this.props.status_message}</div>;
            case 'finished':
                return <div>Finished: {this.props.status_message}</div>;
            default:
                return <div>Unknown status: {this.props.status}</div>;
        }
    }
}

function DropdownMenu(props) {

    // return <div>Testing..... {JSON.stringify(props)}</div>;
    const [anchorEl, setAnchorEl] = React.useState(null);

    const handleClick = event => {
        setAnchorEl(event.currentTarget);
    };

    const handleClose = () => {
        setAnchorEl(null);
    };

    return (
        <div>
        <Button disabled={props.disabled} aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
            {props.label}
        </Button>
        <Menu
            id="simple-menu"
            anchorEl={anchorEl}
            keepMounted
            open={Boolean(anchorEl)}
            onClose={handleClose}
        >
            {
                props.options.map((option) => {
                    return (
                        <MenuItem
                            onClick={() => {handleClose(); props.onSorterSelected(option.sorter);}}
                            key={option.label}>{option.label}
                        </MenuItem>
                    );
                })
            }
        </Menu>
        </div>
    );
}

function isEmpty(obj) {
    return (Object.getOwnPropertyNames(obj).length == 0);
}